docs: Configuring Plugins page intro, page tweaks, and rename (#16534)
eslint/eslint: docs/src/user-guide/configuring/plugins.md

---
title: Plugins & Parsers
layout: doc
eleventyNavigation:
    key: configuring plugins
    parent: configuring
    title: Configuring Plugins & Parsers
    order: 4

---

You can extend ESLint with plugins in a variety of different ways. Plugins can include:

* Custom processors to extract JavaScript code from other kinds of files or preprocess code before linting.
* Custom rules to validate if your code meets a certain expectation, and what to do if it does not meet that expectation.

> * Custom configurations.

Author: added a line

> * Custom configurations.

Reviewer: Looks like perhaps you didn't push that change yet?

> * Custom configurations.

Author: Hmm, on mobile now but will check when back on computer. Sorry if forgot to push!

> * Custom processors to extract JavaScript code from other kinds of files or preprocess code before linting.
> * Custom rules to validates if your code meets a certain expectation, and what to do if it does not meet that expectation.
> * Other custom configuration using ESLint's built-in features.

Reviewer 2: ```suggestion
* Custom processors to extract JavaScript code from other kinds of files or preprocess code before linting.
* Custom rules to validate if your code meets a certain expectation, and what to do if it does not meet that expectation.
* Other custom configurations using ESLint's built-in features.
```

> * Other custom configurations using ESLint's built-in features.

Reviewer 3: ```suggestion
* Custom configurations.
```

There are no configurations before this item, so "Other" sounds confusing. Also, custom configurations can use both built-in features and plugins (e.g., a configuration can enable some rules from the plugin and some built-in rules).

> You can also use custom parsers to convert JavaScript code into an abstract syntax tree for ESLint to evaluate. You might want to add a custom parser if the code you're writing isn't compatible with the ESLint default parser, Espree.

Reviewer 4: ```suggestion
You can also use custom parsers to convert JavaScript code into an abstract syntax tree for ESLint to evaluate. You might want to add a custom parser if the code you've written isn't compatible with the ESLint's default parser Espree.
```

Would this be better?

Reviewer 5: You can also use custom parsers to convert JavaScript code into an abstract syntax tree for ESLint to evaluate. You might want to add a custom parser if the code you've written isn't compatible with ESLint's default parser Espree.

Reviewer 5: ```suggestion
You can also use custom parsers to convert JavaScript code into an abstract syntax tree for ESLint to evaluate. You might want to add a custom parser if the code you've written isn't compatible with ESLint's default parser Espree.
```

Author: ```suggestion
You can also use custom parsers to convert JavaScript code into an abstract syntax tree for ESLint to evaluate. You might want to add a custom parser if your code isn't compatible with ESLint's default parser Espree.
```

thanks i think we can reduce the bloat in this sentence a little more w this change

---

> title: Plugins & Parsers
> layout: doc

Reviewer 3: ```suggestion
title: Plugins & Parsers
```

`layout: doc` is the default now (https://github.com/eslint/eslint/pull/16484), so it isn't necessary here.

---

You can extend ESLint with plugins in a variety of different ways. Plugins can include:

* Custom parsers to convert JavaScript code into an abstract syntax tree for ESLint to evaluate. You might want to add a custom parser if the code you're writing isn't compatible with the ESLint default parser, Espree.
* Custom processors to extract JavaScript code from other kinds of files or preprocess code before linting.
* Custom rules to validates if your code meets a certain expectation, and what to do if it does not meet that expectation.

> * Other custom configuration using ESLint's built in features.

Author: ```suggestion
* Other custom configuration using ESLint's built-in features.
```

}
```

> A scoped packages:

Reviewer 2: ```suggestion
A scoped package:
```

#### Use a plugin

> Rules, environments and configs defined in plugins must be referenced with the following convention:

Author: ```suggestion
Rules, environments, and configurations defined in plugins must be referenced with the following convention:
```

```json
{
    "plugins": ["a-plugin"],
    "processor": "a-plugin/a-processor"
}
```

To specify processors for specific kinds of files, use the combination of the `overrides` key and the `processor` key. For example, the following uses the processor `a-plugin/markdown` for `*.md` files.

```json
{
    "plugins": ["a-plugin"],
    "overrides": [
        {
            "files": ["*.md"],
            "processor": "a-plugin/markdown"
        }
    ]
}
```

Processors may make named code blocks such as `0.js` and `1.js`. ESLint handles such a named code block as a child file of the original file. You can specify additional configurations for named code blocks in the `overrides` section of the config. For example, the following disables the `strict` rule for the named code blocks which end with `.js` in markdown files.

```json
{
    "plugins": ["a-plugin"],
    "overrides": [
        {
            "files": ["*.md"],
            "processor": "a-plugin/markdown"
        },
        {
            "files": ["**/*.md/*.js"],
            "rules": {
                "strict": "off"
            }
        }
    ]
}
```

> ESLint checks the file path of named code blocks then ignores those if any `overrides` entry didn't match the file path. Be sure to add an `overrides` entry if you want to lint named code blocks other than `*.js`.

Reviewer 2: ```suggestion
ESLint checks the file path of named code blocks and then ignore those if any `overrides` entry didn't match the file path. Be sure to add an `overrides` entry if you want to lint named code blocks other than `*.js`.
```

Author: i don't think this change is necessary. and it should be `ignores`, not `ignore`

Reviewer 4: How about `and then ignores`? that sounds good to me?

> ## Configuring Parsers

Author: made singular and removed the "-ing" form on all section headers on the page.

in future PR, should update all section headings across entire docs set to be not "-ing" form of word to align w the pages names in the new IA.

1. It must be a Node module loadable from the config file where the parser is used. Usually, this means you should install the parser package separately using npm.
1. It must conform to the [parser interface](../../developer-guide/working-with-custom-parsers).

Note that even with these compatibilities, there are no guarantees that an external parser works correctly with ESLint. ESLint does not fix bugs related to incompatibilities with other parsers.

To indicate the npm module to use as your parser, specify it using the `parser` option in your `.eslintrc` file. For example, the following specifies to use Esprima instead of Espree:

```json
{
    "parser": "esprima",
    "rules": {
        "semi": "error"
    }
}
```

The following parsers are compatible with ESLint:

* [Esprima](https://www.npmjs.com/package/esprima)
* [@babel/eslint-parser](https://www.npmjs.com/package/@babel/eslint-parser) - A wrapper around the [Babel](https://babeljs.io) parser that makes it compatible with ESLint.
* [@typescript-eslint/parser](https://www.npmjs.com/package/@typescript-eslint/parser) - A parser that converts TypeScript into an ESTree-compatible form so it can be used in ESLint.

> Note that when using a custom parser, the `parserOptions` configuration property is still required for ESLint to work properly with features not in ECMAScript 5 by default. Parsers are all passed `parserOptions` and may or may not use them to determine which features to enable.

Reviewer 4: ```
Parsers are all passed `parserOptions` and may or may not use them to determine which features to enable.
```

I feel this part to be quite confusing 😅

Reviewer: I think this wording is fine and shouldn't hold up this PR. We can continue tweaking later.

    // ...
}
```

### Specifying Processor

> Plugins may provide processors. Processors can extract JavaScript code from other kinds of files, then let ESLint lint the JavaScript code or processors can convert JavaScript code in preprocessing for some purpose.

Author: ```suggestion
Plugins may provide processors. Processors can extract JavaScript code from other kinds of files, then let ESLint lint the JavaScript code. Alternatively, processors can convert JavaScript code during preprocessing.
```
